datastew/embedding.py

```diff
@@ -1,97 +1,178 @@
+        """
+        if max_length <= 0:
+            logging.warning(f"max_length is set to {max_length}, using default value 2048")
+            max_length = 2048
+        
+        sanitized_messages = [self.sanitize(msg) for msg in messages]
+        chunks = [sanitized_messages[i:i + max_length] for i in range(0, len(sanitized_messages), max_length)]
         embeddings = []
-        total_chunks = (len(sanitized_messages) + max_length - 1) // max_length
-        current_chunk = 0
-        for i in range(0, len(sanitized_messages), max_length):
-            current_chunk += 1
-            chunk = sanitized_messages[i:i + max_length]
-            response = openai.Embedding.create(input=chunk, model=self.model_name)
-            embeddings.extend([item["embedding"] for item in response["data"]])
-            logging.info("Processed chunk %d/%d", current_chunk, total_chunks)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:
+            futures = {executor.submit(self._process_chunk, chunk): chunk for chunk in chunks}
+            for future in concurrent.futures.as_completed(futures):
+                try:
+                    embeddings.extend(future.result())
+                except Exception as e:
+                    logging.error(f"Error in processing chunk: {e}")
         return embeddings
     
+    def _process_chunk(self, chunk: List[str]) -> Sequence[Sequence[float]]:
+        """Process a batch of text messages to retrieve embeddings.
+        
+        :param chunk: A list of sanitized messages.
+        :return: A sequence of embedding vectors.
+        """
+        try:
+            response = openai.Embedding.create(input=chunk, model=self.model_name)
+            return [item["embedding"] for item in response["data"]]
+        except Exception as e:
+            logging.error(f"Error processing chunk: {e}")
+            return []
+    
     def get_model_name(self) -> str:
+        """Return the name of the embedding model in use.
+        
+        :return: The model name
+        """
         return self.model_name
 
 
 class MPNetAdapter(EmbeddingModel):
-    def __init__(self, model_name="sentence-transformers/all-mpnet-base-v2"):
-        logging.getLogger().setLevel(logging.INFO)
-        self.model = SentenceTransformer(model_name)
+    def __init__(self, model_name="sentence-transformers/all-mpnet-base-v2", num_threads: int = 4):
+        """Initialize the MPNet adapter with a specified model name and threading settings.
+        
+        :param model_name: The model name for sentence transformers.
+        :param num_threads: The number of CPU threads for inference.
+        """
+        self.model = SentenceTransformer(model_name).to("cuda" if torch.cuda.is_available() else "cpu")
         self.model_name = model_name # For Weaviate
-
-    def get_embedding(self, text: str):
-        logging.info(f"Getting embedding for {text}")
+        torch.set_num_threads(num_threads)
```

Comment:
This calls each time an embedding is created, only needs to be called once for the model

Reply:
It is actually called once per model, the problem is how GitHub reported the changes. It looks like it belongs to get_embedding function whereas they have 60 rows of code in between.